server/store.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+)
+
+// storeInfo contains information about a store.
+type storeInfo struct {
+	store *metapb.Store
+	stats *StoreStatus
+}
+
+func newStoreInfo(store *metapb.Store) *storeInfo {
+	return &storeInfo{
+		store: store,
+		stats: newStoreStatus(),
+	}
+}
+
+func (s *storeInfo) clone() *storeInfo {
+	return &storeInfo{
+		store: proto.Clone(s.store).(*metapb.Store),
+		stats: s.stats.clone(),
+	}
+}
+
+func (s *storeInfo) isUp() bool {
+	return s.store.GetState() == metapb.StoreState_Up
+}
+
+func (s *storeInfo) isOffline() bool {
+	return s.store.GetState() == metapb.StoreState_Offline
+}
+
+func (s *storeInfo) isTombstone() bool {
+	return s.store.GetState() == metapb.StoreState_Tombstone
+}
+
+func (s *storeInfo) downTime() time.Duration {
+	return time.Since(s.stats.LastHeartbeatTS)
+}
+
+func (s *storeInfo) usedRatio() float64 {
+	if s.stats.GetCapacity() == 0 {
+		return 0
+	}
+	return float64(s.stats.GetUsedSize()) / float64(s.stats.GetCapacity())
+}
+
+func (s *storeInfo) leaderRatio() float64 {
+	if s.stats.TotalRegionCount == 0 {
+		return 0
+	}
+	return float64(s.stats.LeaderRegionCount) / float64(s.stats.TotalRegionCount)
+}
+
+// StoreStatus contains information about a store's status.
+type StoreStatus struct {
+	*pdpb.StoreStats
+
+	StartTS           time.Time `json:"start_ts"`
+	LastHeartbeatTS   time.Time `json:"last_heartbeat_ts"`
+	TotalRegionCount  int       `json:"total_region_count"`
+	LeaderRegionCount int       `json:"leader_region_count"`
+}
+
+func newStoreStatus() *StoreStatus {
+	return &StoreStatus{
+		StoreStats: &pdpb.StoreStats{},
+		StartTS:    time.Now(),
+	}
+}
+
+func (s *StoreStatus) clone() *StoreStatus {
+	return &StoreStatus{
+		StoreStats:        proto.Clone(s.StoreStats).(*pdpb.StoreStats),
+		StartTS:           s.StartTS,
+		LastHeartbeatTS:   s.LastHeartbeatTS,
+		TotalRegionCount:  s.TotalRegionCount,
+		LeaderRegionCount: s.LeaderRegionCount,
+	}
+}
+
+func (s *StoreStatus) update(stats *pdpb.StoreStats, totalRegionCount, leaderRegionCount int) {
+	s.StoreStats = stats
+	s.LastHeartbeatTS = time.Now()
+	s.TotalRegionCount = totalRegionCount
+	s.LeaderRegionCount = leaderRegionCount
+}
+
+// GetUptime returns the uptime of the store.
+func (s *StoreStatus) GetUptime() time.Duration {
+	uptime := s.LastHeartbeatTS.Sub(s.StartTS)
```

Comment:
Any where to change the StartTS? 
If the store is down for a long time, what's the up time?